servers/jak/shared/common/main/java/com/couchbase/lite/mobiletest/data/KeypathParser.java

```diff
@@ -0,0 +1,171 @@
+//
+// Copyright (c) 2023 Couchbase, Inc All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package com.couchbase.lite.mobiletest.data;
+
+import androidx.annotation.NonNull;
+
+import com.couchbase.lite.mobiletest.errors.ClientError;
+
+
+public class KeypathParser {
+    private enum CharClass {EOP, DOLLAR, DOT, OBRACKET, CBRACKET, DIGIT, MISC}
+
+    private final String keyPath;
+    private final int n;
+
+    private CharClass curCharClass;
+    private String curChar;
+    private int i = -1;
+
+    public KeypathParser(@NonNull String keyPath) {
+        this.keyPath = keyPath;
+        this.n = keyPath.length();
+    }
+
+    @NonNull
+    public KeyPath parse() {
+        final KeyPath parsedPath = new KeyPath();
+
+        // unless the first path element is an index,
+        // pretend there is an inital dot (introducing a property)
+        final String ch = peekCh();
+        if ("[".equals(ch)) { nextChar(); }
+        else { pushChar("."); }
+
+        while (true) {
+            switch (curCharClass) {
+                case DOT:
+                    nextChar();
+                    parsedPath.addElement(parseProperty());
+                    break;
+                case OBRACKET:
+                    nextChar();
+                    parsedPath.addElement(parseIndex());
+                    break;
+                case EOP:
+                    if (parsedPath.size() < 1) { throw new ClientError(errMsg("Empty keypath")); }
+                    return parsedPath;
+                default:
+                    throw new ClientError(errMsg("Unexpected character " + curChar));
+            }
+        }
+    }
+
+    @NonNull
+    private KeyPath.PathElem.Property parseProperty() {
+        // skip and unescaped leading '$'
+        if (curCharClass == CharClass.DOLLAR) { nextChar(); }
```

Comment:
I believe the current way dollar sign is supposed to be handled is the same way as JSON paths, meaning that dollar sign indicates the root of the object so it would look like `$.foo.bar`.  This just looks like being able to add an arbitrary dollar sign onto the front?

Reply:
Yes.  Fixed.  I sent around a grammar for what I've ended up implementing.  I think it is more in line with your ideas.